### PR TITLE
Add _collections_abc module

### PR DESCRIPTION
**Pull request by CraftSpider**
I'm quite amused: The way typeshed does this, re-exporting from typing, is the reverse of how runtime does it

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Should we put the implementations here and make `collections.abc` `import *` from here? That more closely matches what runtime does but I'm not sure it makes a practical difference.

**Comment by CraftSpider** [comment]
Personally, I'd like to do that to match runtime, but I don't think it matters either way

File: stdlib/3/_collections_abc.pyi [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    Hashable as Hashable,
+    Iterable as Iterable,
+    Iterator as Iterator,
+    Sized as Sized,
+    Callable as Callable,
+    Mapping as Mapping,
+    MutableMapping as MutableMapping,
+    Sequence as Sequence,
+    MutableSequence as MutableSequence,
+    Set as Set,
```
Comment (hauntsaninja) [inline stdlib/3/_collections_abc.pyi:22]: I think this should be AbstractSet?

Comment (hauntsaninja) [inline stdlib/3/_collections_abc.pyi:22]:
```suggestion
    AbstractSet as Set,
```

Comment (CraftSpider) [reply stdlib/3/_collections_abc.pyi:22]: Yep, my mistake. Thanks

File: stdlib/3/collections/__init__.pyi [hunk 2]
```diff
@@ -8,7 +8,7 @@ from typing import (
 # These are exported.
 from . import abc
 
-from typing import (
+from _collections_abc import (
```
Comment (hauntsaninja) [inline stdlib/3/collections/__init__.pyi:11]: Can this just be `from _collections_abc import *`? I guess we should maybe guard it with a `if sys.version_info < (3, 10):` :-)

Comment (CraftSpider) [reply stdlib/3/collections/__init__.pyi:11]: Yeah, it can be. The implementation in collection forwards everything in `__all__`

Comment (CraftSpider) [reply stdlib/3/collections/__init__.pyi:11]: Upon trying to do this, pytype and mypy have started choking on import from collections / collections.abc. Even before, pytype chokes on MutableMapping (and only that one) for a reason I don't know

Comment (hauntsaninja) [reply stdlib/3/collections/__init__.pyi:11]: What's the mypy error?

Comment (CraftSpider) [reply stdlib/3/collections/__init__.pyi:11]:
```
running mypy --python-version 3.8 --strict-optional --no-site-packages --show-traceback --no-implicit-optional --disallow-any-generics --disallow-subclassing-any # with 957 files
third_party\2and3\requests\packages\urllib3\_collections.pyi:2: error: Module 'collections.abc' has no attribute 'MutableMapping'
third_party\2and3\requests\packages\urllib3\_collections.pyi:11: error: Class cannot subclass value of type 'Any'
third_party\2and3\requests\packages\urllib3\_collections.pyi:24: error: Class cannot subclass value of type 'Any'
stdlib\3\shelve.pyi:5: error: Name 'collections.abc.MutableMapping' is not defined
stdlib\3\shelve.pyi:5: error: Class cannot subclass value of type 'Any'
third_party\2and3\markupsafe\__init__.pyi:4: error: Module 'collections' has no attribute 'Mapping'
```

Repeated the same for each of the versions. Changing the imports to `from collections.abc` as relevant doesn't fix the issue.

**Comment by hauntsaninja** [comment]
Oh, right, this was a little mysterious to me, and seems to still be failing. The changes look fairly straightforward, so not sure what's going on. If I had to take a guess — I've seen some weird behaviour surrounding use of `__all__` in stubs and this could well be `__all__`-related; maybe try removing all mentions of `__all__`?

File: stdlib/3/collections/__init__.pyi (hunk 2 above)
Comment (CraftSpider) [reply stdlib/3/collections/__init__.pyi]: Solved the error, it's the same as the issue in #3923 with __all__

**Comment by CraftSpider** [comment]
See my reply above, it is because of `__all__`. Thank you for the help

**Comment by CraftSpider** [comment]
Playing around with pytype, I think it fixes the issue to add `MutableMapping = MutableMapping` after the import in _collections_abc. Is this my fault, or is pytype choking on the re-exports? (ping @rchen152)

**Comment by JukkaL** [comment]
> Should we put the implementations here and make collections.abc import * from here? 

I think that this might result in mypy showing `_collections_abc` in some error messages, so the way these are done currently is arguably a bit better for mypy usability. Also, the current structure is probably more convenient for somebody who just wants to look up how something is defined in the stubs, as they won't have to dig deeper into internal modules.

**Comment by rchen152** [comment]
Sorry for the belated response - we noticed today that pytype seems to choke on re-importing a name without changing it (so `from x import y as z` works, but `from x import y as y` does not), so the CI issue here seems like a pytype bug. Hopefully will be fixed shortly.

**Comment by rchen152** [comment]
It's a little hard to tell since the PR doesn't merge cleanly anymore, but I believe pytype 2020.09.24 fixes the pytype test failure. Let me know if you're still having issues.

**Comment by JelleZijlstra** [comment]
This has some pretty bad merge conflicts now, @CraftSpider do you want to fix them? I think this should be good to go once CI is green.

**Comment by CraftSpider** [comment]
I'll fix the merge conflicts as soon as I have the free time to

**Pull request merged**